Blocks/Persistence/file.h

```diff
@@ -426,117 +412,125 @@ class FilePersister {
       }
     }
 
-    operator bool() const { return valid_; }
+    operator bool() const { return file_persister_impl_; }
 
    private:
-    mutable ScopeOwnedBySomeoneElse<FilePersisterImpl> file_persister_impl_;
-    bool valid_ = true;
+    const Borrowed<FilePersisterImpl> file_persister_impl_;
     const uint64_t begin_;
     const uint64_t end_;
     const std::streampos begin_offset_;
   };
 
-  template <current::locks::MutexLockStatus MLS, typename E, typename US>
-  idxts_t DoPublish(E&& entry, const US us) {
-    current::locks::SmartMutexLockGuard<MLS> lock(file_persister_impl_->mutex_ref);
+  // `TIMESTAMP` can be `std::chrono::microseconds` or `current::time::DefaultTimeArgument`.
+  template <current::locks::MutexLockStatus MLS, typename E, typename TIMESTAMP>
+  idxts_t PersisterPublishImpl(E&& entry, TIMESTAMP&& provided_timestamp) {
+    current::locks::SmartMutexLockGuard<MLS> lock(file_persister_impl_->publish_mutex_ref_);
 
-    end_t iterator = file_persister_impl_->end.load();
-    const auto timestamp = current::time::GetTimestampFromLockedSection(us);
+    end_t iterator = file_persister_impl_->end_.load();
+    const auto timestamp = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(provided_timestamp));
     if (!(timestamp > iterator.head)) {
+#ifdef CURRENT_BUILD_WITH_PARANOIC_RUNTIME_CHECKS
+      std::cerr << "timestamp: " << timestamp.count() << ", iterator.head: " << iterator.head.count() << std::endl;
+#endif  // CURRENT_BUILD_WITH_PARANOIC_RUNTIME_CHECKS
       CURRENT_THROW(ss::InconsistentTimestampException(iterator.head + std::chrono::microseconds(1), timestamp));
     }
 
     iterator.last_entry_us = iterator.head = timestamp;
     const auto current = idxts_t(iterator.next_index, iterator.last_entry_us);
```

Comment:
`current` is confusing/ambiguous here; suggest renaming to `current_index` (or just `index`, because `entry`, not `current_entry`).